zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ZooKeeperUpdatingListener.java

```diff
@@ -30,68 +37,96 @@
  * into the ZooKeeper.
  */
 public class ZooKeeperUpdatingListener extends ServerListenerAdapter {
-    private final String zkConnectionStr;
+
+    /**
+     * Creates a ZooKeeper server listener, which registers server into ZooKeeper.
+     *
+     * <p>If you need a fully customized {@link ZooKeeperUpdatingListener} instance, use
+     * {@link ZooKeeperUpdatingListenerBuilder} instead.
+     *
+     * @param zkConnectionStr ZooKeeper connection string
+     * @param zNodePath       ZooKeeper node path(under which this server will be registered)
+     */
+    public static ZooKeeperUpdatingListener of(String zkConnectionStr, String zNodePath) {
+        return new ZooKeeperUpdatingListenerBuilder(zkConnectionStr, zNodePath).build();
+    }
+
+    private final CuratorFramework client;
     private final String zNodePath;
-    private final int sessionTimeout;
+    private final NodeValueCodec nodeValueCodec;
     @Nullable
     private Endpoint endpoint;
-    @Nullable
-    private ZooKeeperRegistration connector;
+    private final boolean closeClientOnStop;
+
+    ZooKeeperUpdatingListener(CuratorFramework client, String zNodePath, NodeValueCodec nodeValueCodec,
+                              @Nullable Endpoint endpoint, boolean closeClientOnStop) {
+        this.client = requireNonNull(client, "client");
+        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
+        this.nodeValueCodec = requireNonNull(nodeValueCodec, "nodeValueCodec");
+        this.endpoint = endpoint;
+        this.closeClientOnStop = closeClientOnStop;
+    }
 
     /**
-     * A ZooKeeper server listener, used for register server into ZooKeeper.
+     * A ZooKeeper server listener, which registers server into ZooKeeper.
+     *
+     * @deprecated Use {@link ZooKeeperUpdatingListenerBuilder}.
+     *
      * @param zkConnectionStr ZooKeeper connection string
-     * @param zNodePath       ZooKeeper node path(under which this server will registered)
+     * @param zNodePath       ZooKeeper node path(under which this server will be registered)
      * @param sessionTimeout  session timeout
      * @param endpoint        the endpoint of the server being registered
      */
+    @Deprecated
     public ZooKeeperUpdatingListener(String zkConnectionStr, String zNodePath, int sessionTimeout,
                                      Endpoint endpoint) {
-        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
+        checkArgument(!Strings.isNullOrEmpty(zkConnectionStr), "zkConnectionStr");
```

Comment:
Then we don't need this as well because the `CuratorFrameworkFactory` raises the error?

Reply:
... OMG. I see.